nearswap/src/twap.rs

```diff
@@ -0,0 +1,576 @@
+        price1: u128,
+        price2: u128,
+        max_length: usize
+    ) -> usize {
+        let mut o = &self.observations[self.current_idx];
+        if(block_timestamp == o.block_timestamp) {
+            self.observations[self.current_idx] = Observation::transform(o, block_timestamp, price1, price2);
+            return self.current_idx;
+        }
+
+        if self.current_idx + 1 >= max_length {
+            self.pivoted = true;
+        }
+
+        let updated_index: usize = (self.current_idx + 1) % max_length;
+        if updated_index < self.observations.len() {
+            self.observations[updated_index] = Observation::transform(o, block_timestamp, price1, price2);
+        } else {
+            self.observations.push(Observation::transform(o, block_timestamp, price1, price2));
+        }
+
+        self.current_idx = updated_index;
+        return updated_index;
+    }
+
+    /**
+    Pivoted point binary search: searches array which is
+    sorted and rotated from a particular point.
+    Similar to rotated array from a certain pivot point.
+    Parameters:
+    + `max_length`: The maximum length of TWAP array.
+    + `timestamp`: given timestamp.
+    */
+    pub fn binary_search(
+        &self,
+        max_length: usize,
+        block_timestamp: u64,
+    ) -> usize {
+
+        // edge case when all values are less than required
+        if self.observations[self.current_idx].block_timestamp
+            < block_timestamp {
+            panic!("Observation after this timestamp doesn't exist");
+        }
+
+        let mut start: usize = 0;
+        let mut end: usize = self.current_idx + 1;
+
+        let mut mid: usize;
+
+        while start < end {
+            mid = (start + end) / 2;
+            if block_timestamp <= self.observations[mid].block_timestamp {
+                end = mid;
+            } else {
+                start = mid + 1;
+            }
+        }
+
+        if self.pivoted && start == 0 {
+            let res = start;
+            start = self.current_idx + 1;
+            end = max_length;
+
+            while start < end {
+                mid = (start + end) / 2;
+                if block_timestamp <= self.observations[mid].block_timestamp {
+                    end = mid;
+                } else {
+                    start = mid + 1;
+                }
+            }
+            if start >= max_length {
+                start = res;
+            }
+        }
+
+        return start;
+    }
+
+    // function which will calculate mean using Mean enum
+    pub fn calculate_mean(
+        &self,
+        time: Mean,
+        max_length: usize,
+    ) -> (u128, u128) {
+        let time_diff: u64 = match time {
+            Mean::M_1MIN => to_nanoseconds(60), // 1 minute in nanoseconds
+            Mean::M_5MIN => to_nanoseconds(300), // 5 minute in nanoseconds
+            Mean::M_1H => to_nanoseconds(60 * 60),
+            Mean::M_12H => to_nanoseconds(12 * 60 * 60),
```

Comment:
let's create constants for this instead of recalculating it in every call.

Reply:
```
pub const T_1MIN = to_nanoseconds(60);
...
```

Reply:
ok